docs/code-guidelines.md

```diff
@@ -0,0 +1,1141 @@
+        - [Prefer Specifying Container Capacity](#prefer-specifying-container-capacity)
+            - [Specifying Map Capacity Hints](#specifying-map-capacity-hints)
+            - [Specifying Slice Capacity](#specifying-slice-capacity)
+    - [Function Grouping and Ordering](#function-grouping-and-ordering)
+    - [Reduce Nesting](#reduce-nesting)
+
+## Project organization
+
+- /cmd: Main applications for this project.
+    - genaccounts.go
+    - cmd.go
+    - cmd_test.go
+    - testnet.go
+    - main.go
+
+- /build: Packaging and Continuous Integration.
+    - docker stuff
+
+- /test (tests): Additional external test apps and test data.
+    - /e2e
+
+- /API (client): OpenAPI/Swagger specs, JSON schema files, protocol definition files.
+    - /swagger
+
+- /scripts (contrib): Scripts to perform various build, install, analysis, etc operations.
+    - /devtools
+    - /generate_release_note
+    - /scripts
+    - /testnets
+    - /githooks
+
+- /pkg: Library code that's to be reusable.
+    - /address
+    - /genesis
+
+- /internal: Private application and library code.
+    - /ante
+    - /app
+
+- /tools: Supporting tools for this project.
+    - /proto
+
+- /x: Cosmos Modules
+
+
+### Line Length
+
+Avoid uncomfortably long lines. Similarly, don't add line breaks to keep lines short when they are more readable long--for example if they are repetitive. The maximum line length is 120. If your line is over 120 characters, break it;
+
+### Doc Comments
+
+All top-level, exported names should have doc comments, as should non-trivial unexported type or function declarations. See https://go.dev/doc/effective_go#commentary for more information about commentary conventions.
+
+### Declaring Empty Slices
+
+When declaring an empty slice, prefer
+
+```go
+var t []string
+```
+
+over
+
+```go
+t := []string{}
+```
+
+The former declares a nil slice value, while the latter is non-nil but zero-length. They are functionally equivalent—their `len` and `cap` are both zero—but the nil slice is the preferred style.
+
+Note that there are limited circumstances where a non-nil but the zero-length slice is preferred, such as when encoding JSON objects (a `nil` slice encodes to `null`, while `[]string{}` encodes to the JSON array `[]`).
+
+When designing interfaces, avoid distinguishing between a nil slice and a non-nil, zero-length slice, as this can lead to subtle programming errors.
+
+For more discussion about nil in Go see Francesc Campoy's talk [Understanding Nil](https://www.youtube.com/watch?v=ynoY2xz-F8s).
+
+
+### Indent Error Flow
+
+Try to keep the normal code path at a minimal indentation and indent the error handling, dealing with it first. This improves the readability of the code by permitting visual scanning of the normal path quickly. For instance, don't write:
+
+```go
+if err != nil {
+	// error handling
+} else {
+	// normal code
+}
+```
+
+Instead, write:
+
+```go
+if err != nil {
+	// error handling
+	return // or continue, etc.
```

Comment:
why this is bad ? I feel this sometimes makes the code look cleaner

Reply:
It's not so bad. But the code keeps cleaner if we try to keep it more intended to the left instead to the right. for simple cases, it's good, but the code will be too nested for more complicated conditions